src/ToData.purs

```diff
@@ -10,28 +25,151 @@ import Data.BigInt (BigInt)
   toData :: a -> PlutusData
 
+class ConstrIndex (a :: Type) where
+  resolveIndex :: forall s. IsSymbol s => Proxy a -> SProxy s -> BigInt
+
+class ToDataWithIndex a ci where
+  toDataWithIndex :: ConstrIndex ci => Proxy ci -> a -> PlutusData
+
+-- As explained in https://harry.garrood.me/blog/write-your-own-generics/ this
+-- is just a neat pattern that flattens a skewed Product of Products
+class ToDataArgs a where
+  toDataArgs :: a -> Array (PlutusData)
+
+-- Default constructor indices
+class CountedConstrIndex (a :: Type) where
+  makeConstrIndex :: Proxy a -> Int -> Map String Int -> Map String Int
+
+-- Data.Generic.Rep instances
+
+instance toDataWithIndexSum ::
+  ( ConstrIndex ci
+  , ToDataWithIndex l ci
+  , ToDataWithIndex r ci
+  ) =>
+  ToDataWithIndex (G.Sum l r) ci where
+  toDataWithIndex p (G.Inl x) = toDataWithIndex (p :: Proxy ci) x
+  toDataWithIndex p (G.Inr x) = toDataWithIndex (p :: Proxy ci) x
+
+instance toDataWithIndexConstr ::
+  ( IsSymbol ln
+  , ConstrIndex ci
+  , ToDataArgs la
+  ) =>
+  ToDataWithIndex (G.Constructor ln la) ci where
+  toDataWithIndex p (G.Constructor args) = Constr (resolveIndex p (SProxy :: SProxy ln)) (toDataArgs args)
+
+instance toDataArgsNoArguments :: ToDataArgs G.NoArguments where
+  toDataArgs _ = []
+
+instance toDataArgsArgument :: ToData a => ToDataArgs (G.Argument a) where
+  toDataArgs (G.Argument x) = [ toData x ]
+
+instance toDataArgsProduct :: (ToDataArgs a, ToDataArgs b) => ToDataArgs (G.Product a b) where
+  toDataArgs (G.Product x y) = toDataArgs x <> toDataArgs y
+
+instance countedConstrIndexSum ::
+  ( CountedConstrIndex a
+  , CountedConstrIndex b
+  ) =>
+  CountedConstrIndex (G.Sum a b) where
+  makeConstrIndex _ i sym2ix = makeConstrIndex (Proxy :: Proxy b) (i + 1) (makeConstrIndex (Proxy :: Proxy a) i sym2ix)
+
+instance countedConstrIndexConstr :: (IsSymbol ln) => CountedConstrIndex (G.Constructor ln la) where
+  makeConstrIndex _ i sym2ix = (Map.insert (reflectSymbol (SProxy :: SProxy ln)) i sym2ix)
+
+genericToData
+  :: forall a rep. G.Generic a rep => ToData rep => a -> PlutusData
+genericToData = toData <<< G.from
+
+genericToDataWithIndex
+  :: forall a rep. G.Generic a rep => ConstrIndex a => ToDataWithIndex rep a => a -> PlutusData
+genericToDataWithIndex = toDataWithIndex (Proxy :: Proxy a) <<< G.from
+
+defaultMakeConstrIndex :: forall a rep. G.Generic a rep => CountedConstrIndex rep => Proxy a -> Map String Int
+defaultMakeConstrIndex _ = makeConstrIndex (Proxy :: Proxy rep) 0 Map.empty
+
+defaultResolveIndex :: forall a rep s. G.Generic a rep => CountedConstrIndex rep => IsSymbol s => Proxy a -> SProxy s -> BigInt
+defaultResolveIndex _ constrSym =
+  let
+    constrName = reflectSymbol constrSym
+    constrIndex = defaultMakeConstrIndex (Proxy :: Proxy a)
+  in
+    case Map.lookup constrName constrIndex of
+      Just i -> BigInt.fromInt i
+      Nothing -> zero - BigInt.fromInt 1
+
+-- Instances
+
+-- TODO: Remove Day as it's only for demo
```

Comment:
I believe I saw you mention on slack that you've succeeded in getting the generic PS instances to match Plutus. Would you mind adding a module to the `test` directory testing some types against the expected Plutus output? You could move these demo types there as well to test the instances

Reply:
Sorry for the late response! Adding this to the task list!

Reply:
Hi @ngua , I added some general unit tests, however, I think we should postpone testing for haskell compatibility atm as described in the https://github.com/Plutonomicon/cardano-browser-tx/pull/209#issue-1176034885

Wdyt?

Reply:
Sure, we can leave that for another PR (I'll open another issue)